Source/Scene/ImageBasedLighting.js

```diff
@@ -0,0 +1,577 @@
+   *
+   * @private
+   */
+  shouldRegenerateShaders: {
+    get: function () {
+      return this._shouldRegenerateShaders;
+    },
+  },
+
+  /**
+   * Whether or not to use the default spherical harmonic coefficients.
+   *
+   * @memberof ImageBasedLighting.prototype
+   * @type {Boolean}
+   *
+   * @private
+   */
+  useDefaultSphericalHarmonics: {
+    get: function () {
+      return this._useDefaultSphericalHarmonics;
+    },
+  },
+
+  /**
+   * Whether or not the image-based lighting settings use spherical harmonic coefficients.
+   *
+   * @memberof ImageBasedLighting.prototype
+   * @type {Boolean}
+   *
+   * @private
+   */
+  useSphericalHarmonicCoefficients: {
+    get: function () {
+      return (
+        defined(this._sphericalHarmonicCoefficients) ||
+        this._useDefaultSphericalHarmonics
+      );
+    },
+  },
+
+  /**
+   * The texture atlas for the specular environment maps.
+   *
+   * @memberof ImageBasedLighting.prototype
+   * @type {OctahedralProjectedCubeMap}
+   *
+   * @private
+   */
+  specularEnvironmentMapAtlas: {
+    get: function () {
+      return this._specularEnvironmentMapAtlas;
+    },
+  },
+
+  /**
+   * Whether or not to use the default specular environment maps.
+   *
+   * @memberof ImageBasedLighting.prototype
+   * @type {Boolean}
+   *
+   * @private
+   */
+  useDefaultSpecularMaps: {
+    get: function () {
+      return this._useDefaultSpecularMaps;
+    },
+  },
+
+  /**
+   * Whether or not the image-based lighting settings use specular environment maps.
+   *
+   * @memberof ImageBasedLighting.prototype
+   * @type {Boolean}
+   *
+   * @private
+   */
+  useSpecularEnvironmentMaps: {
+    get: function () {
+      return (
+        (defined(this._specularEnvironmentMapAtlas) &&
+          this._specularEnvironmentMapAtlas.ready) ||
+        this._useDefaultSpecularMaps
+      );
+    },
+  },
+});
+
+function createSpecularEnvironmentMapAtlas(imageBasedLighting, context) {
+  if (!OctahedralProjectedCubeMap.isSupported(context)) {
+    return;
+  }
+
+  imageBasedLighting._specularEnvironmentMapAtlas =
+    imageBasedLighting._specularEnvironmentMapAtlas &&
+    imageBasedLighting._specularEnvironmentMapAtlas.destroy();
+
+  if (defined(imageBasedLighting._specularEnvironmentMaps)) {
+    const atlas = new OctahedralProjectedCubeMap(
+      imageBasedLighting._specularEnvironmentMaps
+    );
+    imageBasedLighting._specularEnvironmentMapAtlas = atlas;
+
+    atlas.readyPromise
+      .then(function () {
+        imageBasedLighting._specularEnvironmentMapLoaded = true;
+      })
+      .catch(function (error) {
+        console.error(`Error loading specularEnvironmentMaps: ${error}`);
+      });
+  }
+
+  // Regenerate shaders so they do not use an environment map.
+  // Will be set to true again if there was a new environment map and it is ready.
+  imageBasedLighting._shouldRegenerateShaders = true;
+}
+
+ImageBasedLighting.prototype.update = function (frameState) {
```

Comment:
I didn't see this method called from `ModelExperimental`, does that need to be added?